### PR TITLE
Fix issue(db file Merge issue in windows env):

### LINKED ISSUE #14: Error if open a existing db folder in windows env
Hi, bitcask works very well in Linux env, but I encounted a issue in windows, if the db folder alread existing, Open("my.db") will get "file is accessing by other process" issue, I debuged bitcask and noticed the error is hanppen in bitcask.go line 305 "err = os.Rename(tempdf.Name(), df.Name())".

So I changed (df *Datafile) Close() method both to below and it works:
```
if df.w != nil{
		err = df.w.Close()
	}

	if err == nil && df.r != nil{
		df.r.Close()
	}

	if err == nil && df.ra != nil{
		df.ra.Close()
	}

	return
```

My question is do you think the fixing is OK?

### PR DESCRIPTION
**Pull request by xapanyun**
You will receive an error with the following steps in windows env:
1. Run my program firstly, it will create my.db folder normally.
2. Close my program and restart it again, I will receive an error.

The root cause is bitcask didn't close all opened file handle.

Fixes #14

### CODE REVIEW

File: internal/datafile.go [hunk 1]
```diff
@@ -90,20 +90,24 @@ func (df *Datafile) Name() string {
 	return df.r.Name()
 }
 
-func (df *Datafile) Close() error {
-	if df.w == nil {
-		err := df.ra.Close()
-		if err != nil {
-			return err
+func (df *Datafile) Close() (err error) {
+	if df.w != nil{
```
Comment (prologic) [inline internal/datafile.go]:
```suggestion
	if df.w != nil {
```

File: internal/datafile.go [hunk 2]
```diff
@@ -90,20 +90,24 @@ func (df *Datafile) Name() string {
...
+		err = df.Sync()
+
+		if err == nil{
```
Comment (prologic) [inline internal/datafile.go]:
```suggestion
		if err == nil {
```

File: internal/datafile.go (hunk 1 above)
Comment (xapanyun) [reply internal/datafile.go]: I didn't find a place to put "defer file.Close() " in "Merge" method body.

**Comment by prologic** [comment]
What I'm suggesting is this:

```#!diff
diff --git a/internal/datafile.go b/internal/datafile.go
index 576f928..1738b73 100644
--- a/internal/datafile.go
+++ b/internal/datafile.go
@@ -92,18 +92,17 @@ func (df *Datafile) Name() string {
 
 func (df *Datafile) Close() error {
 	if df.w == nil {
-		err := df.ra.Close()
-		if err != nil {
-			return err
-		}
-		return df.r.Close()
+		defer func() {
+			df.r.Close()
+		}()
+		return df.ra.Close()
 	}
 
-	err := df.Sync()
-	if err != nil {
-		return err
-	}
-	return df.w.Close()
+	defer func() {
+		df.w.Close()
+	}()
+
+	return df.Sync()
 }
 
 func (df *Datafile) Sync() error {
```

**Comment by xapanyun** [comment]
> return df.ra.Close()

If "df.w" is not nil, you will miss that “df.r” closing

**Comment by prologic** [comment]
You are quite right!

I think this makes more sense; and the use of `defer` here makes it a bit more clear:

```#!diff
diff --git a/internal/datafile.go b/internal/datafile.go
index 576f928..ad96637 100644
--- a/internal/datafile.go
+++ b/internal/datafile.go
@@ -91,12 +91,14 @@ func (df *Datafile) Name() string {
 }
 
 func (df *Datafile) Close() error {
+	defer func() {
+		df.ra.Close()
+		df.r.Close()
+	}()
+
+	// Readonly Datafile -- Nothing further to close on the write side
 	if df.w == nil {
-		err := df.ra.Close()
-		if err != nil {
-			return err
-		}
-		return df.r.Close()
+		return nil
 	}
 
 	err := df.Sync()
```

The entire function now looks like this:

```#!go
func (df *Datafile) Close() error {
        defer func() {
                df.ra.Close()
                df.r.Close()
        }()

        // Readonly Datafile -- Nothing further to close on the write side
        if df.w == nil {
                return nil
        }

        err := df.Sync()
        if err != nil {
                return err
        }
        return df.w.Close()
}
```

What do you think? Can ou test this on your Windows machine? All tests pass here for me on macOS and Linux.

**Comment by prologic** [comment]
@xapanyun Let's get this PR merged today? Did you manage to test the deferred-style fix?

**Comment by xapanyun** [comment]
> @xapanyun Let's get this PR merged today? Did you manage to test the deferred-style fix?

Hold on, it seems ok, but let me test it in windows.

**Comment by xapanyun** [comment]
> > @xapanyun Let's get this PR merged today? Did you manage to test the deferred-style fix?
> 
> Hold on, it seems ok, but let me test it in windows.

I already test it in windows, it works well.

**Comment by prologic** [comment]
Perfect; let me review the change one last time...

**Comment by prologic** [comment]
I'd still do another defer for closing `df.w` but this is okay :) Merging! Thank you for filing and fixing this bug!

**Pull request merged**